src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -1029,6 +1034,52 @@ private void WriteNewLine(Span<byte> output)
             output[BytesPending++] = JsonConstants.LineFeed;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteIndentation(Span<byte> buffer, int indentation)
+        {
+            EnsureIndentation();
+
+            if (_indentByte is not null)
+            {
+                JsonWriterHelper.WriteIndentation(buffer, indentation, (byte)_indentByte);
+            }
+            else if (_indentBytes is { Length: > 0 } indentBytes)
+            {
+                JsonWriterHelper.WriteIndentation(buffer, indentation, indentBytes.Span);
+            }
+        }
+
+        private void EnsureIndentation()
+        {
+            Debug.Assert(_options.Indented);
+
+            if (_indentByte is not null || _indentBytes is not null) return;
```

Comment:
This is adding a lot of checks in what is a method called by the hot path. The indent byte analysis should be made at construction time (and ideally only once at the individual `JsonSerializerOptions`/`JsonWriterOptions` instances).

Reply:
You could extract this to a validation method in `JsonWriterHelper.cs` that both validates the input and extracts the indentation byte. It should be called by each of the setter methods you've added and the analysis results passed on to `JsonWriterOptions`/`Utf8JsonWriter` via internal APIs.

Reply:
My intent here is to compute the indentation stuff only when indentation is really needed (and not just enabled).

Reply:
Are there cases where it wouldn't be used? Probably just simple values like `42` or `null` which aren't that common. I don't think it's worth the added checks imposed every time the hot path method is being called.

Reply:
OK. I have moved all the indentation logic into a new `RawIndentation` struct.
BTW if you have a better name, do not hesitate to share...